pkg/pd/api.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package pd
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pingcap/tidb-dashboard/pkg/utils/distro"
+	"github.com/pingcap/tidb-dashboard/pkg/utils/host"
+)
+
+func fetchEndpoints(c *Client) (endpoints map[string]struct{}, err error) {
+	d, err := FetchMembers(c)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints = make(map[string]struct{}, len(d.Members))
+	for _, m := range d.Members {
+		ip, port, _ := host.ParseHostAndPortFromAddressURL(m.ClientUrls[0])
+		endpoints[fmt.Sprintf("%s:%d", ip, port)] = struct{}{}
+	}
+	return
+}
+
+type InfoMembers struct {
+	Count   int          `json:"count"`
+	Members []InfoMember `json:"members"`
+}
+
+type InfoMember struct {
+	GitHash       string   `json:"git_hash"`
+	ClientUrls    []string `json:"client_urls"`
+	DeployPath    string   `json:"deploy_path"`
+	BinaryVersion string   `json:"binary_version"`
+	MemberID      uint64   `json:"member_id"`
+}
+
+func FetchMembers(c *Client) (*InfoMembers, error) {
```

Comment:
This should be called UnsafeFetchMembers, as it bypasses the target address restriction.